d1985r0.md

```diff
@@ -414,6 +423,163 @@ reflection.
 TODO. Concievably simplifies the implementation and makes it far more
 general.
 
+# Covariance and Contravariance
+
+Consider the two concepts:
+
+```cpp
+template <typename T> concept A = true;
+// B subsumes A
+template <typename T> concept B = A<T> && true;
+```
+
+::: cmptable
+
+## Covariance
+```cpp
+auto returns_a() -> is_a;
+auto returns_b() -> is_b;
+auto f() {
+    // OK, requirement less constrained than returned value
+    is_a x = returns_b();
+    // Error, requirement stricter than returned value
+    is_b y = returns_a();
+}
+```
+
+## Contravariance
+```cpp
+template <template <B> typename f>
+using puts_b = void;
+template <template <A> typename f>
+using puts_a = void;
+template <B> using takes_b = void;
+template <A> using takes_a = void;
+using x = puts_b<takes_a>; // ok
+// Error, constraint mismatch (gcc)
+// clang accepts in error
+using w = puts_a<takes_b>;
+```
+
+:::
+
+We are used to the covariant case, but the usage of contravariant cases is
+not as common. The issue with `puts_a<takes_b>` is that `puts_a` requires a
+_more general_ function than one that just accepts `B`s.
+
+We have seen that concept-constrained template parameters behave correctly -
+covariantly on returns, contravariantly on parameters; but do other template
+parts as well?
+
+Let's just replace `is_a` with `auto` and `is_b` with `int`:
+
+```cpp
+template <template <int> typename f> using puts_int = void;
+template <template <auto> typename f> using puts_auto = void;
+template <int> using takes_int = void;
+template <auto> using takes_auto = void;
+using x = puts_int<takes_auto>; // OK
+using w = puts_auto<takes_int>; // OK??? MSVC, GCC and clang all accept
+```
+
+Both co- and contravariance of function pointers is disallowed completely,
+except for virtual functions, where (pointer and reference) covariant return
+types are allowed, but there are no contravariant parameters.
+
+How about parameter packs?
+
+```cpp
+template <template <typename> typename f> using puts_one = void;
+template <template <typename...> typename f> using puts_var = void;
+template <typename> using takes_one = void;
+template <typename...> using takes_var = void;
+using x = puts_one<takes_var>; // Error! ... behaves covariantly!
```

Comment:
Yes, this should be changed. Or, wait, wasn't this already changed between 14 and 17 or something. Yes: https://godbolt.org/z/HMpsGx at least gcc thinks so. If you change the compiler switch to C++17 the errors go away.